packages/ai-jsx/src/batteries/docs.tsx

```diff
@@ -789,3 +806,56 @@ export async function DocsQA<ChunkMetadata extends Jsonifiable = Jsonifiable>(pr
     </ChatCompletion>
   );
 }
+
+const ResultSchema: z.ZodObject<any> = z
+  .object({
+    answer: z.string().describe("The answer to the user's question"),
+    sources: z.array(z.string()).describe('The title or URL of each document used to answer the question'),
+  })
+  .required({ answer: true });
+
+export type QAWithSourcesResult = z.infer<typeof ResultSchema>;
+
+/** A default component for formatting DocsQAWithSources results. */
+function DefaultQAResultFormatter(result: QAWithSourcesResult) {
+  if (result.sources?.length) {
+    return `${result.answer}\n\nSources:\n${result.sources.join('\n')}`;
+  }
+  return result.answer;
+}
+
+/**
+ * Similar to {@link DocsQA}, but encourages the LLM to return sources for its answer.
+ */
+export async function* DocsQAWithSources<ChunkMetadata extends Jsonifiable = Jsonifiable>(
+  props: DocsQAProps<ChunkMetadata>,
+  { render }: AI.ComponentContext
+) {
+  const chunks = await props.corpus.search(props.question, { limit: props.chunkLimit });
+  const chunkFormatter: (props: { doc: ScoredChunk<ChunkMetadata> }) => Node = props.chunkFormatter ?? DefaultFormatter;
+  const resultFormatter: (result: QAWithSourcesResult) => Node = props.resultFormatter ?? DefaultQAResultFormatter;
+
+  const stringifiedResult = (
+    <JsonChatCompletion schema={ResultSchema}>
+      <SystemMessage>
+        You are a trained question answerer. Answer questions truthfully, using only the document excerpts below. Do not
+        use any other knowledge you have about the world. If you don't know how to answer the question, just say "I
+        don't know." Here are the relevant document excerpts you have been given:
+        {chunks.map((chunk) => chunkFormatter({ doc: chunk }))}
+        And here is the question you must answer:
+      </SystemMessage>
+      <UserMessage>{props.question}</UserMessage>
+    </JsonChatCompletion>
+  );
+
+  const frames = render(stringifiedResult);
+  for await (const frame of frames) {
+    try {
+      yield resultFormatter(ResultSchema.parse(JSON.parse(frame)));
+    } catch (e) {
+      // The intermediate result isn't valid yet. Return it as is instead.
```

Comment:
`JsonChatCompletion` should always emit valid JSON, because as the result is streaming in, it "untruncates" it. (Thanks @farzadab!)

That said, I did observe GPT-4 failing to emit valid JSON approximately 3% of the time. But in that case, it's fully borked and will never recover, so we wouldn't say "isn't valid yet".

Also, `resultFormatter` should tolerate a partial JSON object, because of how the JSON is streamed in – to encode that, I'd change this to:

```ts
QAWithSourcesResult = Partial<z.infer<typeof ResultSchema>>;
```

Reply:
Also, this could be a good place for a `logger.trace` or `logger.debug`. (`logger` is on `AI.ComponentContext`.)

Reply:
Yup, if you pass in the schema, the result should always match that too (sorry for a period there, this didn't work correctly, but it should now and if you see it working otherwise let me know to fix it).

Nick: What does `Partial` do?

Also I'm generally a bit confused about what type of logging should go where since I don't know what happens in terminal vs react vs next. So my logging is usually pretty sloppy.

Reply:
> Also I'm generally a bit confused about what type of logging should go where since I don't know what happens in terminal vs react vs next. So my logging is usually pretty sloppy.

Hmm, can you tell me more about what you're trying to figure out here?

Reply:
Re: partial https://www.perplexity.ai/search/5d43f745-8205-468a-b9fe-64664383328d?s=c